### PR TITLE
Add support for persistence constructors

### PR DESCRIPTION
**Pull request by nmervaillie**
See https://jira.spring.io/browse/DATAGRAPH-1056

This PR is designed to work with this OGM branch : https://github.com/neo4j/neo4j-ogm/tree/feature/persistence-ctors
tracked by https://github.com/neo4j/neo4j-ogm/pull/452

- [X] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [X] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATAGRAPH).
- [X] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [X] You submit test cases (unit or integration tests) that back your changes.
- [X] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

### CODE REVIEW

File: spring-data-neo4j/pom.xml [hunk 1]
```diff
@@ -194,6 +194,80 @@
...
+			<scope>test</scope>
+		</dependency>
+
+	</dependencies>
+
+	<build>
+		<plugins>
+			<plugin>
+				<groupId>org.jetbrains.kotlin</groupId>
+				<artifactId>kotlin-maven-plugin</artifactId>
```
Comment (frant-hartm) [inline spring-data-neo4j/pom.xml]: 1. Do we now compile everything (also java) by the kotlin compiler?
2. Is that safe? 
3. Is this how other SD projects do it?

File: spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/config/Neo4jMappingContextFactoryBean.java [hunk 2]
```diff
@@ -52,16 +59,28 @@ public void setApplicationContext(ApplicationContext applicationContext) throws
...
 		Neo4jMappingContext context = new Neo4jMappingContext(sessionFactory.metaData());
 		context.initialize();
 
+		ConversionService conversionService = null;
+		try {
+			conversionService = beanFactory.getBean(ConversionService.class);
+		} catch (NoSuchBeanDefinitionException e) {
+			LOG.debug("Unable to find a conversion service to use for entity instantiation, using none");
+		}
+		sessionFactory.setEntityInstantiator(getEntityInstantiator(context, conversionService));
```
Comment (frant-hartm) [inline spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/config/Neo4jMappingContextFactoryBean.java]: `getEntityInstantiator(context, conversionService)` may be called with `conversionService = null` - is this ok?

File: spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/config/OgmEntityInstantiatorAdapter.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+			Assert.notNull(propertyValues, "Properties cannot be null");
+			this.propertyValues = propertyValues;
+		}
+
+		@SuppressWarnings({"unchecked"})
+		@Override
+		@Nullable
+		public Object getParameterValue(PreferredConstructor.Parameter parameter) {
+			Object value = propertyValues.get(parameter.getName());
+			if (value == null || conversionService == null) {
```
Comment (frant-hartm) [inline spring-data-neo4j/src/main/java/org/springframework/data/neo4j/repository/config/OgmEntityInstantiatorAdapter.java]: I guess this answers my previous question.

File: src/main/asciidoc/reference/neo4j-repositories.adoc [hunk 4]
```diff
@@ -663,5 +663,41 @@ Delete graph entities::
...
+
+It means that SDN will try to find the best constructor to instantiate entities.
+Constructor discovery works as follow :
+
+- look for explicit @PersistenceConstructor annotation on constructor
+- if not found, look for the default constructor
+- if not found, look at any specific constructor defined by the entity.
+
+Values for constructor parameters are determined by looking at parameters names (not parameter types).
+For values to be filled in correctly, the parameter names have to match the entity attributes names.
```
Comment (frant-hartm) [inline src/main/asciidoc/reference/neo4j-repositories.adoc]: 1. Doesn't this mean domain classes need to be compiled with `-parameters` flag? Should be mentioned here.
1. How does the test code end up with it? I don't see it anywhere..

File: spring-data-neo4j/pom.xml (hunk 1 above)
Comment (nmervaillie) [reply spring-data-neo4j/pom.xml]: I just applied the suggestions from Intellij. Looks a lot liek what's in the kotlin official docs.
I presume (maybe wrong) kotlin does not reimplement a full java compiler but delegates to javac.

File: src/main/asciidoc/reference/neo4j-repositories.adoc (hunk 4 above)
Comment (nmervaillie) [reply src/main/asciidoc/reference/neo4j-repositories.adoc]: 1. No. Actually when the code in not compiled with -parameters flag, Spring Data falls back to byte code parsing to get argument names.

**Comment by nmervaillie** [comment]
merged 2af6a45

**Pull request closed without merging**